File: animation.go
package main

import (
	"path/filepath"

	"github.com/qeedquan/go-media/sdl"
)

const delaySet = 3

type Animation struct {
	loop         bool
	finished     bool
	name         string
	knownAs      string
	currentFrame int
	frames       []*sdl.Texture
	delay        int
}

func newAnimation(spriteName, filename string, loop bool, knownAs string) *Animation {
	name := filepath.Join(spriteName, filename, "sprite.dat")
	s := newScanner(name)
	defer s.close()

	var frames []*sdl.Texture
	var numFrames int

	s.scan(&numFrames)
	for i := 0; i < numFrames; i++ {
		var str string

		s.scan(&str)
		textureFile := filepath.Join(spriteName, filename, str)
		frames = append(frames, loadTexture(textureFile))
	}

	return &Animation{
		name:    filename,
		loop:    loop,
		delay:   delaySet,
		knownAs: knownAs,
		frames:  frames,
	}
}

func (a *Animation) getKnown() string {
	return a.knownAs
}

func (a *Animation) reset() {
	a.currentFrame = 0
	a.finished = false
}

func (a *Animation) getFrameN(fnum int) *sdl.Texture {
	if fnum >= 0 && fnum < len(a.frames) {
		return a.frames[fnum]
	}
	return a.frames[0]
}

func (a *Animation) getFrame() *sdl.Texture {
	if a.currentFrame == len(a.frames)-1 && a.loop {
		a.currentFrame = 0
	}
	if a.currentFrame == len(a.frames)-1 && !a.loop {
		a.finished = true
	}
	if a.currentFrame < len(a.frames)-1 {
		a.delay--
		if a.delay == 0 {
			a.currentFrame++
			a.delay = delaySet
		}
	}
	return a.frames[a.currentFrame]
}

File: orig/hhh-2008-03-29/Animation.cc
// Loads and stores an animation sequence
//
#include "Animation.h"
#include <string>
#include <fstream>
#include <iostream>

#define delay_set 3

Animation::Animation(char* spritename, char* filename, bool loop, char* knownAs){
	this->currentframe = 0;
	this->name = filename;
	this->loop = loop;
	this->finished = false;
	this->delay = delay_set;
	numframes = 0;

	string sprite(spritename);
	string fullpath(filename);

	this->knownAs = knownAs;
	//printf("Loading sprite at %s/%s/sprite.dat\n",spritename,filename);
	#ifdef WIN32
	fullpath = sprite + "\\" + fullpath + "\\sprite.dat";
	#else
	fullpath = DATA_INSTALL_DIR "/" + sprite + "/" + fullpath + "/sprite.dat";
	#endif
	//printf("Opening %s \n",fullpath.c_str());
	ifstream br(fullpath.c_str());
	char buffer[30];
	if(!br.is_open()){
		printf("Could not open spritefile!\n");
	} else {
		// File opened OK, get numframes followed by list
		//printf("File Opened OK\n");
		br.getline(buffer,30);
		numframes = atoi(buffer);
		//printf("Reading %d lines\n",numframes);
		int i = 0;
		for(i = 0; i < numframes; i++){
			br.getline(buffer,30);
			//printf("Loading image %s \n", buffer);
			string filepath(filename);
			filepath = "/" + sprite + "/" + filepath + "/" + buffer;
			string dir(DATA_INSTALL_DIR);
			dir = dir + "/" + filepath;
			//printf("Filepath: %s\n",filepath.c_str());
			SDL_Surface* s;
			s = SDL_DisplayFormatAlpha(IMG_Load(dir.c_str()));
			//SDL_SetColorKey(s, SDL_SRCCOLORKEY,
				//SDL_MapRGB(s->format,0x00,0xFF,0xFF));

			frames.push_back(s);
		}		
		numframes = frames.size();
		br.close();
	}
	//printf("Finished with Animation constructor: %d frames\n",numframes);
};

char* Animation::getKnown(){
	return knownAs;
};

void Animation::reset(){
	currentframe=0;
	finished=false;
};

SDL_Surface* Animation::getFrame(int fnum){
	//cout << knownAs << " > " << fnum << "\n";
	if(fnum>=0 && fnum<numframes-1){
		return frames[fnum];
	} else {
		return frames[0];
	}
};

SDL_Surface* Animation::getFrame(){
	if(currentframe==numframes-1 && loop==true){
		currentframe = 0;
	}
	if(currentframe==numframes-1 && loop==false){
		currentframe = currentframe;
		finished = true;
	} 
	if(currentframe<numframes-1){
		delay--;
		if(delay==0){
			currentframe++;
			delay = delay_set;
		}
	}
	//printf("Return frame %d\n",currentframe);
	return frames[currentframe];
};

Animation::~Animation(){
	int current = 0;
	for(current=0; current<numframes; current++){
		//cout << "Deleting Animation Frame\n";
		SDL_FreeSurface(frames[current]);
	}
};
